packages/core/src/Queue.js

```diff
@@ -1,64 +1,98 @@
-/** @license MIT License (c) copyright 2010-2016 original author or authors */
-/** @author Brian Cavalier */
-/** @author John Hann */
+/** @license MIT License (c) copyright 2010 original author or authors */
 
-// Based on https://github.com/petkaantonov/deque
+// Based on https://github.com/invertase/denque
 
 export default class Queue {
-  constructor (capPow2 = 32) {
-    this._capacity = capPow2
-    this._length = 0
+  constructor () {
     this._head = 0
+    this._tail = 0
+    this._capacityMask = 0x3
+    this._list = new Array(4)
   }
 
   push (x) {
-    const len = this._length
-    this._checkCapacity(len + 1)
+    const tail = this._tail
+    this._list[tail] = x
+    this._tail = (tail + 1) & this._capacityMask
+    if (this._tail === this._head) {
+      this._growArray()
+    }
 
-    const i = (this._head + len) & (this._capacity - 1)
-    this[i] = x
-    this._length = len + 1
+    if (this._head < this._tail) {
+      return this._tail - this._head
+    } else {
+      return this._capacityMask + 1 - (this._head - this._tail)
+    }
   }
 
   shift () {
     const head = this._head
-    const x = this[head]
+    if (head === this._tail) {
+      return undefined
+    }
+
+    const x = this._list[head]
+    this._list[head] = undefined
+    this._head = (head + 1) & this._capacityMask
+    if (head < 2 && this._tail > 10000 && this._tail <= this._list.length >>> 2) {
+      this._shrinkArray()
+    }
 
-    this[head] = void 0
-    this._head = (head + 1) & (this._capacity - 1)
-    this._length--
     return x
   }
 
   isEmpty () {
-    return this._length === 0
+    return this._head === this._tail
   }
 
   length () {
-    return this._length
+    if (this._head === this._tail) {
+      return 0
+    } else if (this._head < this._tail) {
+      return this._tail - this._head
+    } else {
+      return this._capacityMask + 1 - (this._head - this._tail)
+    }
   }
 
-  _checkCapacity (size) {
-    if (this._capacity < size) {
-      this._ensureCapacity(this._capacity << 1)
+  _growArray () {
+    if (this._head) {
+      // copy existing data, head to end, then beginning to tail.
+      this._list = this._copyArray(true)
+      this._head = 0
     }
+
+    // head is at 0 and array is now full, safe to extend
+    this._tail = this._list.length
+
+    this._list.length *= 2
+    this._capacityMask = (this._capacityMask << 1) | 1
   }
 
-  _ensureCapacity (capacity) {
-    const oldCapacity = this._capacity
-    this._capacity = capacity
+  _shrinkArray () {
+    this._list.length >>>= 1
+    this._capacityMask >>>= 1
+  }
 
-    const last = this._head + this._length
+  _copyArray (fullCopy) {
```

Comment:
The underscore indicates a private method. `_copyArray` is only being called from _one_ place within the class with `fullCopy` set to `true`. I see no reason why L83 is making a check that _always_ will be _true_.

Reply:
See https://github.com/mostjs/core/pull/235#discussion_r184567162.  I didn't want to change the denque code, if possible.  Happy to discuss other options, tho.

Reply:
What are other options if you don’t want to change the code?

Reply:
Another option is to add denque as a dependency.  I feel like copying the code is a better solution in this case (see https://github.com/mostjs/core/pull/235#discussion_r185921969).

I do want to make sure we add proper attribution, though.  There's no license in their repo, though :/  We could open an issue to confirm they're ok with this approach as long as we include attribution.

Reply:
the package.json and badge both say MIT if that counts for anything

Reply:
In relation to this comment, https://github.com/mostjs/core/pull/235#discussion_r185966918, I also think we should optimize this function.

Reply:
I still don’t see any reason to include the `fullCopy` parameter for the reasons mentioned in [my first comment](https://github.com/mostjs/core/pull/235/files/1685078eb2c627bc31649a37ab93516170cf25ef#r184538393).